mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala

```diff
@@ -211,14 +213,17 @@ private object IDFModel {
     val n = v.size
     v match {
       case SparseVector(size, indices, values) =>
+        val newElements = new ArrayBuffer[(Int, Double)]
         val nnz = indices.size
-        val newValues = new Array[Double](nnz)
         var k = 0
         while (k < nnz) {
-          newValues(k) = values(k) * idf(indices(k))
+          val newValue = values(k) * idf(indices(k))
```

Comment:
The existing code is fine since the idf vector is always dense which has constant time lookup. In your change, that will strongly impact the performance, so I don't think it should belong to a separate PR.